samcli/lib/sync/flows/rest_api_sync_flow.py

```diff
@@ -48,15 +51,156 @@ def __init__(
     def set_up(self) -> None:
         super().set_up()
         self._api_client = self._boto_client("apigateway")
+        self._api_physical_id = self.get_physical_id(self._api_identifier)
 
     def sync(self) -> None:
-        api_physical_id = self.get_physical_id(self._api_identifier)
         if self._definition_uri is None:
             raise MissingLocalDefinition(ResourceIdentifier(self._api_identifier), "DefinitionUri")
-        LOG.debug("%sTrying to put RestAPI through client", self.log_prefix)
-        response = self._api_client.put_rest_api(restApiId=api_physical_id, mode="overwrite", body=self._swagger_body)
-        LOG.debug("%sPut RestApi Result: %s", self.log_prefix, response)
+
+        self._update_api()
+        new_dep_id = self._create_deployment()
+        stages = self._collect_stages()
+        prev_dep_ids = self._update_stages(stages, new_dep_id)
+        self._delete_deployments(prev_dep_ids)
+
+    def _update_api(self) -> None:
+        """
+        Update the API content
+        """
+        LOG.debug("%sTrying to update RestAPI through client", self.log_prefix)
+        response_put = cast(
+            Dict,
+            self._api_client.put_rest_api(restApiId=self._api_physical_id, mode="overwrite", body=self._swagger_body),
+        )
+        LOG.debug("%sPut RestApi Result: %s", self.log_prefix, response_put)
+
+    def _create_deployment(self) -> Optional[str]:
+        """
+        Create a deployment using the updated API and record the created deployment ID
+
+        Returns
+        ----------
+        Optional[str]: The newly created deployment ID
+        """
+        LOG.debug("%sTrying to create a deployment through client", self.log_prefix)
+        response_dep = cast(
+            Dict, self._api_client.create_deployment(restApiId=self._api_physical_id, description="Created by SAM Sync")
+        )
+        new_dep_id = response_dep.get("id")
+        LOG.debug("%sCreate Deployment Result: %s", self.log_prefix, response_dep)
+        return new_dep_id
+
+    def _collect_stages(self) -> Set[str]:
+        """
+        Collect all stages needed to be updated
+
+        Returns
+        ----------
+        Set[str]: The set of stage names to be updated
+        """
+        # Get the stage name associated with the previous deployment and update stage
+        # Stage needs to be flushed so that new changes will be visible immediately
+        api_resource = get_resource_by_id(self._stacks, ResourceIdentifier(self._api_identifier))
+        stage_resources = get_resource_ids_by_type(self._stacks, "AWS::ApiGateway::Stage")
+
+        stages = set()
+        # If it is a SAM resource, get the StageName property
+        if api_resource:
+            if api_resource.get("Type") == "AWS::Serverless::Api":
+                # The customer defined stage name
+                stage_name = api_resource.get("Properties").get("StageName")  # type: ignore
+                stages.add(cast(str, stage_name))
```

Comment:
Would there be a case where customer did not defined a `StageName`? If so we should be checking whether the value of that property is not `None`.

Reply:
Translation would fail if the property StageName is not defined. All serverless api resources are required to have this field. But in the rare case that customer never ran any other sam commands before this and directly call code sync, this might be a problem. So adding a check also makes sense.